### PR TITLE
Adds user authentication to movies comment page.

### PR DESCRIPTION
**Pull request by vlop11**
Uses the Users API in App Engine to authenticate users with their Google account. Users must be logged in in order to post a comment.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -36,24 +39,28 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     
+    // keeping this just in case I need a filter for future reference
     // Filter movieFilter = new FilterPredicate("movie", FilterOperator.EQUAL, movie);
     // Query query = new Query("Comment").setFilter(movieFilter).addSort("timestamp", SortDirection.DESCENDING);
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It's better to take these out to keep your code clean.

Sometimes I save these in a notes google doc or on a Google Keep note.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -65,11 +72,13 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     String comment = request.getParameter("comment");
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: use constant strings for these

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+            responseMap.put("Link", logoutUrl);
+        } else {
+            responseMap.put("Status", "false");
+
+            String loginUrl = userService.createLoginURL("/movies.html");
+            responseMap.put("Link", loginUrl);
+        }
+
+        response.setContentType("application/json;");
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:40]: Use constants for these too

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -36,24 +39,28 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:41]: There's a comment above that says "TODO:". I think you've made the changes to handle comment data right? If so, you can take out the "TODO:" part of the comment.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -36,24 +39,28 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    // keeping this just in case I need a filter for future reference
     // Filter movieFilter = new FilterPredicate("movie", FilterOperator.EQUAL, movie);
     // Query query = new Query("Comment").setFilter(movieFilter).addSort("timestamp", SortDirection.DESCENDING);
+    
     Query query = new Query(movie).addSort("timestamp", SortDirection.DESCENDING);
     PreparedQuery pq = datastore.prepare(query);
     List<Entity> resultsList = pq.asList(FetchOptions.Builder.withLimit(maxComments));
 
-    ArrayList<String> comments = new ArrayList<>();
+    ArrayList<HashMap<String, String>> comments = new ArrayList<>();
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Could you explain why you need a list of hashmaps here? seems a bit overkill, but maybe i don't understand what you're trying to do (: Seems like each map would just contain a username and a comment? In that case, you can make this a list of pairs.
e.g. List<Pair<String, String>>

Also, use ImmutableList here too :)

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        response.getWriter().println(convertToJSON(responseMap));
+    }
+
+    private String convertToJSON(HashMap<String, String> map) {
+        Gson gson = new Gson();
+        
+        return gson.toJson(map);
+    }
+}
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java]: Leave a newline at the end of each file

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (vlop11) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I tried using Pair but it doesn't come with Java 8 :( apparently it comes with javafx, but when I run it, it doesn't import it saying that "javafx.util doesn't exist." Would it be better to stick with the HashMap or install javafx?

Comment (sporkert) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Oh wow I didn't realize it's not in Java 8! I think a better option would to make a class or enum to store the two strings together, but I won't block merging on it for now (:

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * logged into their Google account. If the user is logged
+ * in, it provides a logout link. If the user is not logged
+ * in, it provides a login link. 
+ */
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        HashMap<String, String> responseMap = new HashMap<>();
```
Comment (sporkert) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:24]: It seems like you're just putting true/false as the value, so consider making this <String, boolean>

**Review by sporkert — APPROVED** [review]
Added some comments but not blocking on it so that you can move forward with the project (:

**Pull request merged**